android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -462,29 +462,102 @@ public static boolean executeHardwareKeystroke(
   }
 
   /**
-   * Handle the globe key longpress action. Currently just for INAPP only
+   * Handle the globe key action on the lock screen
+   * @param context
+   */
+  private static void doGlobeKeyLockscreenAction(Context context) {
+    // Override globe key action if screen is locked:
+    // 1. Switch to next Keyman keyboard (no menu)
+    // 2. When all the Keyman keyboards have been cycled through, advance to the next system keyboard
+    if (sysKbStartingIndexOnLockScreen == getCurrentKeyboardIndex(context)) {
+      // all the Keyman keyboards have been cycled through
+      advanceToNextInputMode();
+    } else {
+      if (sysKbStartingIndexOnLockScreen == -1) {
+        // Initialize the system keyboard starting index while the screen is locked
+        sysKbStartingIndexOnLockScreen = getCurrentKeyboardIndex(context);
+      }
+      switchToNextKeyboard(context);
+    }
+  }
+
+  /**
+   * Handle the globe key longpress action.
    * @param context
    * @param keyboard
    */
   private static void doGlobeKeyLongpressAction(Context context, KeyboardType keyboard) {
-    if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP && InAppKeyboard != null) {
-      if (InAppKeyboard.keyboardPickerEnabled) {
-        showKeyboardPicker(context, keyboard);
+    boolean keyboardPickerEnabled = false;
+    if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP) {
+      if (InAppKeyboard == null) {
+        return;
       }
+      keyboardPickerEnabled = InAppKeyboard.keyboardPickerEnabled;
+    } else if (keyboard == KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+      if (SystemKeyboard == null) {
+        return;
+      }
+      keyboardPickerEnabled = SystemKeyboard.keyboardPickerEnabled;
+    }
```

Comment:
In what scenarios can we have keyboard != KEYBOARD_TYPE_SYSTEM and keyboard != KEYBOARD_TYPE_INAPP?
```suggestion
    } else {
      // assertion failure?
      return;
    }
```